File: htmlvault/scanner.py
"""
RETIRED — HTMLVault has moved to a Chrome extension as its only supported form.

This Python CLI (`htmlvault serve` / `htmlvault scan`) is kept in the repo
for historical reference only. It is no longer installed by pyproject.toml,
no longer published to PyPI, and no longer maintained. The logic here has
been reimplemented in `extension/` as MV3-compliant JavaScript.

If you are a non-technical user: please install the Chrome extension
(see README.md for instructions).

If you want to run the old CLI, revert to an earlier git tag (<= v0.1.0).
The entire original source is preserved below as a docstring so it can be
read, copied, or revived, but it will NOT execute as a module.
"""

_RETIRED_ORIGINAL_SOURCE = r"""
\"\"\"Directory scanner — find user-created HTML files.\"\"\"

import os
from typing import List, Dict

DEFAULT_EXCLUDES = {
    "node_modules", ".venv", "venv", ".git", ".next",
    "dist", "build", "out", "site-packages", "__pycache__",
    ".claude", ".trash", ".cache", ".tox", ".mypy_cache",
    ".pytest_cache", "egg-info", ".worktrees",
}

# Files that are clearly not user content
SKIP_FILENAMES = {
    "LICENSES.chromium.html",
}

MIN_FILE_SIZE = 500  # bytes — skip empty shells


def scan(directories: List[str], extra_excludes: List[str] = None) -> List[Dict]:
    \"\"\"Scan directories for user-created HTML files.

    Returns list of dicts with keys:
        path (str): absolute path
        relpath (str): relative to scan root
        name (str): filename without extension
        folder (str): parent directory relative to scan root
        mtime (float): modification timestamp
        size (int): file size in bytes
    \"\"\"
    excludes = DEFAULT_EXCLUDES.copy()
    if extra_excludes:
        excludes.update(extra_excludes)

    results = []
    for root_dir in directories:
        root_dir = os.path.expanduser(root_dir)
        root_dir = os.path.abspath(root_dir)
        if not os.path.isdir(root_dir):
            continue
        _scan_dir(root_dir, root_dir, excludes, results)

    # Sort by mtime descending (newest first)
    results.sort(key=lambda f: f["mtime"], reverse=True)
    return results


def _scan_dir(
    current: str, root: str, excludes: set, results: list
) -> None:
    \"\"\"Recursively scan a directory.\"\"\"
    dirs = []
    try:
        # Use as context manager so the directory handle is released even if
        # iteration raises.
        with os.scandir(current) as entries:
            for entry in entries:
                if entry.name.startswith("."):
                    continue  # skip all hidden dirs/files
                try:
                    is_dir = entry.is_dir(follow_symlinks=False)
                except OSError:
                    continue
                if is_dir:
                    if entry.name in excludes:
                        continue
                    if any(ex in entry.name for ex in ("egg-info",)):
                        continue
                    dirs.append(entry.path)
                    continue
                try:
                    is_file = entry.is_file()
                except OSError:
                    continue
                if is_file and entry.name.endswith(".html"):
                    if entry.name in SKIP_FILENAMES:
                        continue
                    try:
                        stat = entry.stat()
                    except OSError:
                        continue
                    if stat.st_size < MIN_FILE_SIZE:
                        continue

                    relpath = os.path.relpath(entry.path, root)
                    folder = os.path.dirname(relpath)
                    name = os.path.splitext(entry.name)[0]

                    results.append({
                        "path": entry.path,
                        "relpath": relpath,
                        "name": name,
                        "folder": folder if folder else ".",
                        "mtime": stat.st_mtime,
                        "size": stat.st_size,
                    })
    except (PermissionError, OSError):
        return

    for d in sorted(dirs):
        _scan_dir(d, root, excludes, results)


def format_size(size_bytes: int) -> str:
    \"\"\"Human-readable file size.\"\"\"
    if size_bytes < 1024:
        return f"{size_bytes} B"
    elif size_bytes < 1024 * 1024:
        return f"{size_bytes // 1024} KB"
    else:
        return f"{size_bytes / (1024 * 1024):.1f} MB"
"""

File: extension/scanner.js
// Directory scanner — recursive scan via File System Access API.
// Ported from htmlvault/scanner.py.

const DEFAULT_EXCLUDES = new Set([
  "node_modules", ".venv", "venv", ".git", ".next",
  "dist", "build", "out", "site-packages", "__pycache__",
  ".claude", ".trash", ".cache", ".tox", ".mypy_cache",
  ".pytest_cache", "egg-info", ".worktrees",
]);

const SKIP_FILENAMES = new Set([
  "LICENSES.chromium.html",
]);

const MIN_FILE_SIZE = 500; // bytes

/**
 * Scan a directory handle recursively for HTML files.
 * @param {FileSystemDirectoryHandle} dirHandle - root directory
 * @param {string} dirName - display name for this root
 * @param {function} onProgress - optional callback(count) for progress updates
 * @returns {Promise<Array>} - list of file objects
 */
async function scanDirectory(dirHandle, dirName, onProgress) {
  const results = [];
  await _scanRecursive(dirHandle, "", dirName, results, onProgress);
  // Sort by mtime descending (newest first)
  results.sort((a, b) => b.mtime - a.mtime);
  return results;
}

async function _scanRecursive(dirHandle, relPrefix, rootName, results, onProgress) {
  const subdirs = [];
  try {
    for await (const [name, entry] of dirHandle.entries()) {
    // Skip hidden entries (dot files/dirs)
    if (name.startsWith(".")) continue;

    if (entry.kind === "directory") {
      if (DEFAULT_EXCLUDES.has(name)) continue;
      if (name.includes("egg-info")) continue;
      subdirs.push({ name, handle: entry });
    } else if (entry.kind === "file" && (name.endsWith(".html") || name.endsWith(".htm"))) {
      if (SKIP_FILENAMES.has(name)) continue;

      let file;
      try {
        file = await entry.getFile();
      } catch (e) {
        continue;
      }

      if (file.size < MIN_FILE_SIZE) continue;

      const relpath = relPrefix ? relPrefix + "/" + name : name;
      const folder = relPrefix || ".";
      const baseName = name.replace(/\.html?$/i, "");

      results.push({
        name: baseName,
        relpath: relpath,
        folder: folder,
        mtime: file.lastModified / 1000, // convert ms to seconds
        size: file.size,
        dirName: rootName,
        _fileHandle: entry, // keep handle for reading content later
      });

      if (onProgress) onProgress(results.length);
    }
    }
  } catch (e) {
    // Permission revoked mid-traversal or handle invalid — return what we have.
    return;
  }

  // Recurse into subdirectories (sorted for deterministic order)
  subdirs.sort((a, b) => a.name.localeCompare(b.name));
  for (const sub of subdirs) {
    const subPrefix = relPrefix ? relPrefix + "/" + sub.name : sub.name;
    await _scanRecursive(sub.handle, subPrefix, rootName, results, onProgress);
  }
}

/**
 * Read HTML content from a file handle (for categorization or preview).
 * Reads up to maxBytes (default 200KB, matching Python version).
 * @param {FileSystemFileHandle} fileHandle
 * @param {number} maxBytes
 * @returns {Promise<string>}
 */
async function readFileContent(fileHandle, maxBytes) {
  maxBytes = maxBytes || 200000;
  try {
    const file = await fileHandle.getFile();
    if (file.size <= maxBytes) {
      return await file.text();
    }
    // Read partial — slice the file
    const blob = file.slice(0, maxBytes);
    return await blob.text();
  } catch (e) {
    return "";
  }
}

/**
 * Read full file content as blob URL (for iframe preview or opening).
 * @param {FileSystemFileHandle} fileHandle
 * @returns {Promise<string>} blob URL
 */
async function readFileAsBlobUrl(fileHandle) {
  try {
    const file = await fileHandle.getFile();
    return URL.createObjectURL(file);
  } catch (e) {
    return "";
  }
}

/**
 * Human-readable file size.
 */
function formatSize(bytes) {
  if (bytes < 1024) return bytes + " B";
  if (bytes < 1024 * 1024) return Math.floor(bytes / 1024) + " KB";
  return (bytes / (1024 * 1024)).toFixed(1) + " MB";
}
